File: Go_Files/dominance.go
package main

import "fmt"
import "game/game"


func main () {	
	
	g1 := game.Game{
		GameName: "TestGame",
		RowPlayer: "Rose",
		ColumnPlayer: "Colin",
		Matrix: []game.Rose {
			[]int{10, 10, 10},	
			[]int{5, 10, 11},
			[]int{10, 10, 10},
		},		
	}

	//g1.Show()//Ejemplo de llamada al metodo.

	b, err := isDominant(&g1,"col", 0, 1)

	//Este es el equivalente a un try catch en Go:
	if err == nil {
		fmt.Printf("%v\n", b)
	} else {
		fmt.Printf("there was an error: '%v'.\n", err)
	}
}

//Esta funcion toma un juego (su puntero)
//Los indices de dos estrategias y
//El tipo de jugador para determinar si la primera domina la segunda.
func isDominant(
	gamePointer *game.Game, 
	playerType string, 
	strategyIndex1 int, 
	strategyIndex2 int) (bool, error) {

		var oneStriclyBetter = false

		if playerType == "row" {
			
			strategy1 := (*gamePointer).Matrix[strategyIndex1]
			strategy2 := (*gamePointer).Matrix[strategyIndex2]
			
			for i := range strategy1 {
				if !(strategy1[i] >= strategy2[i]) {
					return false, nil
				}
				if strategy1[i] > strategy2[i] {
					oneStriclyBetter = true
				}
			}
			return oneStriclyBetter, nil 
		}


		if playerType == "col" {

			strategy1 := []int{}			
			strategy2 := []int{}

			for _, row := range (*gamePointer).Matrix {
				strategy1 = append(strategy1, row[strategyIndex1])
				strategy2 = append(strategy2, row[strategyIndex2])
			}

			for i := range strategy1 {
				if !(strategy1[i] <= strategy2[i]) {
					return false, nil
				}
				if strategy1[i] < strategy2[i] {
					oneStriclyBetter = true
				}
			}
			return oneStriclyBetter, nil 
		}

		//En go los errores son tratados fuera de las funciones
		return false, fmt.Errorf("Opcion no valida")

}

File: Python_Files/dominance.py
from GameClass import Game

g23 = Game(
			"Game 2.2",
			"Rose",
			"Colin",
			[
				#C D E
				[1,2,2],#A
				[1,1,2],#B
				[1,1,1],#C
				[2,1,0],#D
			]
	)

g21 = Game(
			"Game 2.1",
			"Nina",
			"Juan",
			[
				#A B C D E
				[1,1,1,2,2],#A
				[2,1,1,1,2],#B
				[2,2,1,1,1],#C
				[2,2,2,1,0]#D
			]
	)

# def is_stricly_better(
# 	one_stricly_better,
# 	game,
# 	strategy_index_1,
# 	strategy_index_2,
# 	player_type):

# 	player_name = game.row_name if player_type == "row" else game.column_name

# 	if not(one_stricly_better):
# 		print(f'La estrategia {player_name}[{strategy_index_1}] NO domina a la estrategia {player_name}[{strategy_index_2}] en {game.name}.')
# 	else:
# 		print(f'La estrategia {player_name}[{strategy_index_1}] DOMINA a la estrategia {player_name}[{strategy_index_2}] en {game.name}.')


#Esta funcion determina si una estrategia (strategy_1) de un jugador es dominante sobre otra (strategy_2)
#las estrategias se espesifican por los indices en la matriz del juego
def is_dominant(game, player_type, strategy_index_1, strategy_index_2):

	one_stricly_better = False
	
	if player_type == "row":

		strategy_1 = game.matrix[strategy_index_1]
		strategy_2 = game.matrix[strategy_index_2]

		for i in range(len(strategy_1)):
			if not(strategy_1[i] >= strategy_2[i]):
				#print(f'La estrategia {game.row_name}[{strategy_index_1}] NO domina a la estrategia {game.row_name}[{strategy_index_2}] en {game.name}.')
				return False
			
			if strategy_1[i] > strategy_2[i]:
				one_stricly_better = True

		# is_stricly_better(
		# 	one_stricly_better,
		# 	game,
		# 	strategy_index_1,
		# 	strategy_index_2,
		# 	player_type,
		# )

		return one_stricly_better

	if player_type == "column":

		strategy_1 = []
		strategy_2 = []

		for row in game.matrix:
			strategy_1.append(row[strategy_index_1])
			strategy_2.append(row[strategy_index_2])

		for i in range(len(strategy_1)):
			if not(strategy_1[i] <= strategy_2[i]):
				#print(f'La estrategia {game.column_name}[{strategy_index_1}] NO domina a la estrategia {game.column_name}[{strategy_index_2}] en {game.name}.')
				return False

			if strategy_1[i] < strategy_2[i]:
				one_stricly_better = True


		# is_stricly_better(
		# 	one_stricly_better,
		# 	game,
		# 	strategy_index_1,
		# 	strategy_index_2,
		# 	player_type,
		# )

		return one_stricly_better

# print(is_dominant(g23, "row", 0, 3))
# print(is_dominant(g21, "column", 2, 0))
# print(is_dominant(g21, "column", 2, 1))
